src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,376 @@
+            // orderedTokenNumber:          0         1         2
+            for (var orderedEditNumber = 0; orderedEditNumber < editIndices.Length; orderedEditNumber++)
+            {
+                var orderedTokenNumber = editIndices[orderedEditNumber];
+                var initialEditKind = indexToEditKinds[orderedTokenNumber];
+
+                if (initialEditKind == SemanticTokenEditKind.Update)
+                {
+                    orderedEditNumber = AddUpdateEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndices, orderedEditNumber,
+                        groupedSemanticToken: newGroupedSemanticTokens[orderedTokenNumber],
+                        editStartPosition: orderedTokenNumber * 5);
+                }
+                else if (initialEditKind == SemanticTokenEditKind.Insert)
+                {
+                    orderedEditNumber = AddInsertionEdit(
+                        newGroupedSemanticTokens, indexToEditKinds, semanticTokensEdits, editIndices, orderedEditNumber,
+                        groupedSemanticToken: newGroupedSemanticTokens[orderedTokenNumber],
+                        editStartPosition: orderedTokenNumber * 5);
+                }
+                else
+                {
+                    Contract.ThrowIfFalse(initialEditKind == SemanticTokenEditKind.Delete);
+                    orderedEditNumber = AddDeletionEdit(
+                        indexToEditKinds, semanticTokensEdits, editIndices, orderedEditNumber,
+                        editStartPosition: orderedTokenNumber * 5);
+                }
+            }
+
+            return semanticTokensEdits.ToArray();
+
+            // Local functions
+            static int AddUpdateEdit(
+                SemanticToken[] newGroupedSemanticTokens,
+                Dictionary<int, SemanticTokenEditKind> indexToEditKinds,
+                ArrayBuilder<SemanticTokensEdit> semanticTokensEdits,
+                int[] editIndices,
+                int orderedEditNumber,
+                SemanticToken groupedSemanticToken,
+                int editStartPosition)
+            {
+                var _ = ArrayBuilder<int>.GetInstance(out var tokensToInsert);
+                groupedSemanticToken.AddToEnd(tokensToInsert);
+
+                // For simplicitly, we only allow an "update" (i.e. a dual insertion/deletion) to be
+                // combined with other updates.
+                var updatedOrderedEditNumber = GetUpdatedOrderedEditNumber(
+                    SemanticTokenEditKind.Update, indexToEditKinds, editIndices, orderedEditNumber);
+
+                var deleteCount = 5 + 5 * (updatedOrderedEditNumber - orderedEditNumber);
+                for (var i = 1; i <= updatedOrderedEditNumber - orderedEditNumber; i++)
+                {
+                    newGroupedSemanticTokens[editIndices[orderedEditNumber + i]].AddToEnd(tokensToInsert);
+                }
```

Comment:
question.  if you started from`i = 0` would this be simpler as you could remove the `groupedSemanticToken.AddToEnd(tokensToInsert)` line at the top of hte method?  (you could even then not pass in groupedSemanticToken).  Or am i misreading this?

Reply:
Nope, you're right. I think I was thinking of your suggestion at one point, but it involve re-looking up things in arrays, which maybe is fine because it's inexpensive. I'll change it.